devito/operators.py

```diff
@@ -10,23 +11,123 @@ def __init__(self, subs, nt, shape, spc_border, time_order, forward, dtype):
         for param in self.output_params:
-            self.params[param.name] = param
+            self.params[param.label] = param
             self.propagator.add_devito_param(param)
+        self.symbol_to_data = {}
+        for param in self.input_params+self.output_params:
+            self.symbol_to_data[param.name] = param
         self.propagator.subs = self.subs
         self.propagator.stencils, self.propagator.stencil_args = zip(*self.stencils)
 
-    def apply(self):
+    def apply(self, debug=False):
+        if debug:
+            return self.apply_python()
         f = self.get_callable()
         for param in self.input_params:
             param.initialize()
         args = [param.data for param in self.input_params + self.output_params]
         f(*args)
         return tuple([param.data for param in self.output_params])
 
+    def apply_python(self):
+        self.run_python()
+        return tuple([param.data for param in self.output_params])
+
+    def find_free_terms(self, expr):
+        free_terms = []
+        for term in expr.args:
+            if isinstance(term, Indexed):
+                free_terms.append(term)
+            else:
+                free_terms += self.find_free_terms(term)
+        return free_terms
+
+    def symbol_to_var(self, term, ti, indices=[]):
+        arr = self.symbol_to_data[str(term.base.label)].data
+        if len(self.shape) == 2:
+            space_dims = symbols("x z")
+        else:
+            space_dims = symbols("x y z")
+        num_ind = []
+        for ind in term.indices:
+            ind = ind.subs({symbols("t"): ti}).subs(tuple(zip(space_dims, indices)))
+            num_ind.append(ind)
+        return (arr, tuple(num_ind))
+
+    def expr_to_lambda(self, expr_arr):
+        lambdas = []
+        for expr in expr_arr:
+            terms = self.find_free_terms(expr.rhs)
+            term_symbols = [symbols("i%d" % i) for i in range(len(terms))]
+            # Substitute IndexedBase references to simple variables
+            # lambdify doesn't support IndexedBase references in expressions
+            expr_to_lambda = expr.rhs.subs(dict(zip(terms, term_symbols)))
+            lambdified = lambdify(term_symbols, expr_to_lambda)
+            lambdas.append((lambdified, terms))
+        return lambdas
+
+    def run_python(self):
+        time_loop_limits = self.propagator.time_loop_limits
+        time_loop_lambdas_b = self.expr_to_lambda(self.propagator.time_loop_stencils_b)
+        time_loop_lambdas_a = self.expr_to_lambda(self.propagator.time_loop_stencils_a)
+        stencils = []
+        for stencil, args in self.stencils:
+            stencil = stencil.subs(dict(zip(self.subs, args)))
+            stencils.append(stencil)
+        stencil_lambdas = self.expr_to_lambda(stencils)
+        for ti in range(*time_loop_limits):
+            # Run time loop stencils before space loop
+            for lams, expr in zip(time_loop_lambdas_b, self.propagator.time_loop_stencils_b):
```

Comment:
This loop is not exercised by the test added in this PR. I'm assuming this is used by more complicated operators (Forward, Adjoint, Born, etc)?